cpp/include/cuspatial/experimental/detail/pairwise_point_in_polygon.cuh

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuspatial/error.hpp>
+#include <cuspatial/experimental/detail/is_point_in_polygon_kernel.cuh>
+#include <cuspatial/traits.hpp>
+#include <cuspatial/vec_2d.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+
+#include <thrust/memory.h>
+
+#include <iterator>
+#include <type_traits>
+
+namespace cuspatial {
+namespace detail {
+
+template <class Cart2dItA,
+          class Cart2dItB,
+          class OffsetIteratorA,
+          class OffsetIteratorB,
+          class OutputIt,
+          class Cart2dItADiffType = typename std::iterator_traits<Cart2dItA>::difference_type,
+          class Cart2dItBDiffType = typename std::iterator_traits<Cart2dItB>::difference_type,
+          class OffsetItADiffType = typename std::iterator_traits<OffsetIteratorA>::difference_type,
+          class OffsetItBDiffType = typename std::iterator_traits<OffsetIteratorB>::difference_type>
+__global__ void pairwise_point_in_polygon_kernel(Cart2dItA test_points_first,
```

Comment:
Is pairwise PiP going to be a common case for "Big Data"? I suspect that the common case will be N points, M polys, where N >> M.  In that case, pairwise won't work. And the parallelism model used here won't be efficient for pairwise cases unless N is very large.

Reply:
Pairwise PIP is always N points and N polys?

Reply:
Yes, that's my point. Unless N is large enough, it will underutilize the GPU.  I just couldn't see any existing use of geopandas having a very large number of polygons.  Hence this would be a case where mimicking geopandas is questionable.

Reply:
I think a common case with more data would be pairwise *points* in polygon, where we test multiple points inside each polygon. And in this case, it might be a bad idea to use one thread per pair rather than one thread per point.

Reply:
Eventually this PR will be refactored with the new range objects. And hopefully we should be able to get multipoints/multipolys for free. For now it's probably ok to leave it as is?

Reply:
I agree that multipoint->polygon is probably the more commonly used API. But it's probably a stretch for what this PR is trying to achieve, especially we are only looking to support a subset of binary predicates which is only single-single geometry related.

Reply:
I wrote this while working on `polygon.contains(points)` and as I moved on to linestring and polygons I remembered that pairwise doesn't apply for that case. I'm not sure of the usefulness of this PR, except for the colinearity test which I could easily separate out.

Reply:
Doing the M x N point-in-polygon seems wasteful for doing `polygon.contains(points)` but adding the code complexity may not be worth the inefficiency trade-off.

Reply:
`polygon.contains(points)` presumably checks all points inside a single polygon, so it doesn't matter if it's MxN since M == 1. Or am I missing something?

Reply:
The other use case is `Mpolygons.contains(Mpoints)`. In the pairwise implementation this tests each point $i$ for membership in each polygon $i$. In our existing implementation this tests all points $0..M$ for membership in each polygon $0..M$, then for this algorithm discards all the results except for $M_i,i$. The pairwise operation is $O(M)$ and the all-pairs operation is $O(M^2)$, lots more expensive. I don't know how common this use case is, or if the pairwise implementation will be sufficiently faster than the quadtree implementation mess with. I'm still researching that.

Reply:
OK, thanks.